### PR TITLE
Fix: NaN value making infinite loop inside GraphEdit's `NOTIFICATION_DRAW`

### LINKED ISSUE #70147: `GraphEdit.set_zoom_step` fills entire ram
### Godot version

3.6.beta.custom_build. d79699041

### System information

Ubuntu 22.10 - AMD RX 470, Gnome shell 43 X11

### Issue description

```
func _ready():
	var temp_variable19694 = GraphEdit.new()
	add_child(temp_variable19694)
	temp_variable19694.set_zoom_step(-NAN)
	temp_variable19694._zoom_plus()
```
will start to fill ram ~1GB/s
```
0x00000000047071d7 in CowData<RasterizerCanvas::Item::Command*>::set (this=0x7232a80, p_index=82156841, p_elem=@0x7fffffffc970: 0x1908009c0) at ./core/cowdata.h:140
140	./core/cowdata.h: Directory not empty.
(gdb) backtrace
#0  0x00000000047071d7 in CowData<RasterizerCanvas::Item::Command*>::set (this=0x7232a80, 
    p_index=82156841, p_elem=@0x7fffffffc970: 0x1908009c0) at ./core/cowdata.h:140
#1  0x00000000047045be in Vector<RasterizerCanvas::Item::Command*>::set (this=0x7232a78, 
    p_index=82156841, p_elem=@0x7fffffffc970: 0x1908009c0) at ./core/vector.h:85
#2  0x0000000004702e61 in Vector<RasterizerCanvas::Item::Command*>::push_back (this=0x7232a78, 
    p_elem=0x1908009c0) at ./core/vector.h:190
#3  0x00000000046faa0c in VisualServerCanvas::canvas_item_add_line (this=0x72d2900, p_item=..., 
    p_from=..., p_to=..., p_color=..., p_width=1, p_antialiased=false)
    at servers/visual/visual_server_canvas.cpp:506
#4  0x00000000045919af in VisualServerRaster::canvas_item_add_line (this=0x74dac10, arg1=..., 
    arg2=..., arg3=..., arg4=..., arg5=1, arg6=false) at servers/visual/visual_server_raster.h:700
#5  0x00000000045f393f in VisualServerWrapMT::canvas_item_add_line (this=0x7524e00, p1=..., p2=..., 
    p3=..., p4=..., p5=1, p6=false) at servers/visual/visual_server_wrap_mt.h:601
#6  0x0000000003ce9685 in CanvasItem::draw_line (this=0x8bd0f40, p_from=..., p_to=..., p_color=..., 
    p_width=1, p_antialiased=false) at scene/2d/canvas_item.cpp:705
#7  0x00000000038770e7 in GraphEdit::_notification (this=0x8bd0f40, p_what=30)
    at scene/gui/graph_edit.cpp:513
#8  0x000000000388adbc in GraphEdit::_notificationv (this=0x8bd0f40, p_notification=30, 
    p_reversed=false) at scene/gui/graph_edit.h:97
#9  0x0000000004891d91 in Object::notification (this=0x8bd0f40, p_notification=30, p_reversed=false)
    at core/object.cpp:927
#10 0x0000000003ce78da in CanvasItem::_update_callback (this=0x8bd0f40) at scene/2d/canvas_item.cpp:455
#11 0x000000000189af51 in MethodBind0::call (this=0x81f02e0, p_object=0x8bd0f40, p_args=0x0, 
    p_arg_count=0, r_error=...) at ./core/method_bind.gen.inc:59
#12 0x0000000004891c52 in Object::call (this=0x8bd0f40, p_method=..., p_args=0x0, p_argcount=0, 
    r_error=...) at core/object.cpp:918
#13 0x0000000004887fb3 in MessageQueue::_call_function (this=0x70f30d0, p_target=0x8bd0f40, 
    p_func=..., p_args=0x7fffe657d068, p_argcount=0, p_show_error=false) at core/message_queue.cpp:241
#14 0x00000000048882bd in MessageQueue::flush (this=0x70f30d0) at core/message_queue.cpp:284
#15 0x000000000376d2f1 in SceneTree::iteration (this=0x7202740, p_time=0.0166666675)
    at scene/main/scene_tree.cpp:591
#16 0x0000000001810124 in Main::iteration () at main/main.cpp:2309
#17 0x00000000017ce5df in OS_X11::run (this=0x7fffffffd2d0) at platform/x11/os_x11.cpp:4035
#18 0x00000000017b844a in main (argc=2, argv=0x7fffffffdc48) at platform/x11/godot_x11.cpp:59

```

### Steps to reproduce

Above

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by stmSi**
Fix: #70147

### CODE REVIEW

File: scene/gui/control.cpp [hunk 1]
```diff
@@ -1383,6 +1383,7 @@ void Control::_set_size(const Size2 &p_size) {
 }
 
 void Control::set_size(const Size2 &p_size, bool p_keep_offsets) {
+	ERR_FAIL_COND(isnan(p_size.x) || isnan(p_size.y));
```
Comment (YuriSizov) [inline scene/gui/control.cpp:1386]: Is this a relevant change?

Comment (stmSi) [reply scene/gui/control.cpp:1386]: yes.. `temp_variable512.set_size(-NAN, -NAN)` crashes.

Comment (stmSi) [reply scene/gui/control.cpp:1386]:
```
func _ready() -> void:
	var temp_variable512 = GraphEdit.new()
	add_child(temp_variable512)
	temp_variable512.set_size(Vector2(-NAN, -NAN))
	temp_variable512.set_zoom(temp_variable512.get_zoom() * temp_variable512.get_zoom_step())
```
Yep It crashes...

Comment (stmSi) [reply scene/gui/control.cpp:1386]: There is no exposed `_zoom_minus` or `_zoom_plus` in godot4... so i replace them with this `temp_variable512.set_zoom(temp_variable512.get_zoom() * temp_variable512.get_zoom_step())`

Comment (YuriSizov) [reply scene/gui/control.cpp:1387]: Yes, I get that setting the size crashes, but it's not specific to GraphEdit, it's a method of Control. And the report doesn't even mention it. It mentions `set_begin` instead, which is also a method of Control. You can fix it in the same PR, I just find it kind of random to patch only that specific method of Control. I think either all relevant methods need to have these guards, or none.

Comment (stmSi) [reply scene/gui/control.cpp:1387]: `Hmm.. may be we could do wrapper function to guard without touching the `Control` node?

Comment (YuriSizov) [reply scene/gui/control.cpp:1387]: Well, my point was that we should add guards to Control methods, but it shouldn't be one arbitrary method. There are several sizing related methods that can be guarded.

Comment (stmSi) [reply scene/gui/control.cpp:1387]: Oh.. I see I will look into it.

**Comment by akien-mga** [comment]
Should likely use `is_finite()` to cover both nan and infinity.

**Comment by stmSi** [comment]
Sorry for the late reply. here is the changes. I couldn't find exact `is_finite()` but `isfinite()` did the trick.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**